book-content/chapters/16-the-utils-folder.md

```diff
@@ -0,0 +1,1527 @@
+
+### Solution 5: Combining Generic Types and Functions
+
+Here's the starting point of our `safeFunction`:
+
+```typescript
+type PromiseFunc = () => Promise<any>;
+
+const safeFunction = (func: PromiseFunc) => async () => {
+  try {
+    const result = await func();
+    return result;
+  } catch (e) {
+    if (e instanceof Error) {
+      return e;
+    }
+    throw e;
+  }
+};
+```
+
+The first thing we'll do is update the `PromiseFunc` type to be a generic type. We'll call the type parameter `TResult` to represent the type of the value returned by the promise, and and it to the return type of the function:
+
+```typescript
+type PromiseFunc<TResult> = () => Promise<TResult>;
+```
+
+With this update, we now need to update the `PromiseFunc` in the `safeFunction` to include the type argument:
+
+```typescript
+const safeFunction =
+  <TResult>(func: PromiseFunc<TResult>) =>
+  async () => {
+    try {
+      const result = await func();
+      return result;
+    } catch (e) {
+      if (e instanceof Error) {
+        return e;
+      }
+      throw e;
+    }
+  };
+```
+
+With these changes in place, when we hover over the `safeFunction` call in the first test, we can see that the type argument is inferred as `number` as expected:
+
+```typescript
+it("should return an error if the function throws", async () => {
+  const func = safeFunction(async () => {
+    if (Math.random() > 0.5) {
+      throw new Error("Something went wrong");
+    }
+    return 123;
+  });
+  ...
+
+// hovering over safeFunction shows:
+const safeFunction: <number>(func: PromiseFunc<number>) => Promise<() => Promise<number | Error>>
+```
+
+The other tests pass as well.
+
+Whatever we pass into `safeFunction` will be inferred as the type argument for `PromiseFunc`. This is because the type argument is being inferred _inside_ the generic function.
+
+This combination of generic functions and generic types can make your generic functions a lot easier to read.
+
+### Solution 6: Multiple Type Arguments in a Generic Function
+
+Here's how `PromiseFunc` is currently defined:
+
+```typescript
+type PromiseFunc<TResult> = (...args: any[]) => Promise<TResult>;
+```
+
+The first thing to do is figure out the types of the arguments being passed in. Currently, they're set to one value, but they need to be different based on the type of function being passed in.
+
+Instead of having `args` be of type `any[]`, we want to spread in all of the `args` and capture the entire array.
+
+To do this, we'll update the type to be `TArgs`. Since `args` needs to be an array, we'll say that `TArgs extends any[]`. Note that this doesn't mean that `TArgs` will be typed as `any`, but rather that it will accept any kind of array:
+
+```typescript
+type PromiseFunc<TArgs extends any[], TResult> = (
+  ...args: TArgs
+) => Promise<TResult>;
+```
+
+You might have tried this with `unknown[]` - but `any[]` is the only thing that works in this scenario.
```

Comment:
I'm not sure what you are saying here, `unknown[]` works just fine here: [TS playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBACgTgewLYEsDOEBiBXAdgYwB4AVAQTgHM0oIAPYCXAE2rwGtcEB3XAbQF0ANFGIAlCGmwAbYAD4oAXigAKAFBQoAOm0BDSmgBcI8lVUBKRfPjJ0EEuMkzZAblWr8CXGmBQ0OgGZYePjAKJ6K6lAkJtR0DMysuBzcfEIiDtJyyv7BRtaoGDgE0frCYhKZshYKspE6aCAEKtqaelRGZPrV8gDekRrAcCBQfRpjUB5ePnAVMopQOlw6KD45BMotbWhmruMaM8DYcLhQM47Au2MAvhM6wPgAFioQFqN7KP7PUChTOgQQCE+AFE4Ig4K9+nsDkcThBLuMrpCBg9EFwaPCoIiNFdXO5PN4oGt8PM-IEiiEwrhlPVGsTlN0RpEPioALJ3B6aOB-JjIelQeQABk0AFYIeNgCjuFBcBA0SCwcoAEQAZWQEAlPwoUC4jB8XEQuAoip2kSxp3VMKgAEYAEwAZlcVxNQA)